compiler/src/dotty/tools/repl/ReplDriver.scala

```diff
@@ -209,20 +209,32 @@ class ReplDriver(settings: Array[String],
         /* complete = */ false  // if true adds space when completing
       )
     }
-    implicit val state = newRun(state0)
-    compiler
-      .typeCheck(expr, errorsAllowed = true)
-      .map { tree =>
-        val file = SourceFile.virtual("<completions>", expr, maybeIncomplete = true)
-        val unit = CompilationUnit(file)(using state.context)
-        unit.tpdTree = tree
-        given Context = state.context.fresh.setCompilationUnit(unit)
-        val srcPos = SourcePosition(file, Span(cursor))
-        val (_, completions) = Completion.completions(srcPos)
-        completions.map(_.label).distinct.map(makeCandidate)
+
+    if expr.startsWith(":") then
+      ParseResult.commands.collect {
+        // If expr is only : then we return the commands with : since jline
```

Comment:
I think this strange behaviour is because of how our REPL-specific scala parser works. Here `expr` is the entire line where the completion happens, which corresponds to `line` from `JLineTerminal#Parser#ParsedLine`. But it's only `word` from the same class that gets replaced during completion and because of how our parser works `word` always corresponds to a single token, which is `:` if we don't type anything more but if we do then it's the previous token while the current one is an alphanumeric identifier that follows.
I think the proper solution would be to to move this part of completion logic to `JLineTerminal#Parser#parse` and match on the entire `input` to see if it's a prefix of an existing REPL command and then fallback to parsing actual scala code.

Reply:
Gotcha, I'll look at adding this there 👍🏼